### PR TITLE
WIP: Transaction Pool Layer

### PR DESCRIPTION
**Pull request by maflcko**
This implements a layer around an immutable tx pool. The layer can be seen as a temporary throw-away shell that provides the same interface as `CTxMemPool`. Its primary purpose right now is to be passed into ATMP while testing acceptance of several (potentially depending) transaction and then to be discarded.

One use case could be to determine if smart contracts that are set up with multiple txs would be accepted by current consensus and policy rules.

In the future it could be extended to support recursive wrapping of layers or a way to commit changes that happened in the layer to the underlying pool or layer. Furthermore, it could be extended to be revivable after changes to the underlying layer happened. (As opposed to be single-use)

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Currently based on some other commits, which probably should be reviewed and merged first.

File: src/policy/rbf.h [hunk 1]
```diff
@@ -23,6 +23,6 @@ bool SignalsOptInRBF(const CTransaction &tx);
 // according to BIP 125
 // This involves checking sequence numbers of the transaction, as well
 // as the sequence numbers of all in-mempool ancestors.
-RBFTransactionState IsRBFOptIn(const CTransaction &tx, CTxMemPool &pool) EXCLUSIVE_LOCKS_REQUIRED(pool.cs);
+RBFTransactionState IsRBFOptIn(const CTransaction &tx,const CTxMemPool &pool) EXCLUSIVE_LOCKS_REQUIRED(pool.cs);
```
Comment (promag) [inline src/policy/rbf.h]: Commit "Stacked Transaction Pool"

nit, add space after `,`.

File: src/txmempool.cpp [hunk 2]
```diff
@@ -289,19 +332,27 @@ void CTxMemPool::UpdateForRemoveFromMempool(const setEntries &entriesToRemove, b
...
+    for (const auto& removeIt : entriesToRemove) {
+        pool.UpdateChildrenForRemoval(removeIt);
     }
 }
+void CTxMemPool::UpdateForRemoveFromMempool(const setEntries& entriesToRemove, bool updateDescendants)
+{
+    return ::UpdateForRemoveFromMempool(*this, entriesToRemove, updateDescendants);
+}
 
+void CTxMemPool::Layer::  UpdateForRemoveFromMempool(const setEntries&entriesToRemove, bool updateDescendants)
```
Comment (promag) [inline src/txmempool.cpp]: Commit "Stacked Transaction Pool"

nit, remove extra spaces.

**Review by promag — COMMENTED** [review]
Last commit is overwhelming. Could prefix with `wip: ` until dependencies are merged?

File: src/txmempool.cpp [hunk 3]
```diff
@@ -865,38 +993,96 @@ const CTransaction* CTxMemPool::GetConflictTx(const COutPoint& prevout) const
...
     return boost::optional<txiter>{};
 }
 
-CTxMemPool::setEntries CTxMemPool::GetIterSet(const std::set<uint256>& hashes) const
+boost::optional<CTxMemPool::Layer::txiter> CTxMemPool::Layer::GetIter(const uint256& txid) const
 {
-    CTxMemPool::setEntries ret;
+    const auto& it_cache = m_cache_added.find(txid);
+    if (it_cache != m_cache_added.end()) {
+        CTxMemPool::Layer::txiter{it_cache};
```
Comment (practicalswift) [inline src/txmempool.cpp]: Ownerless expression?

**Comment by practicalswift** [comment]
This PR fails to compile with thread safety analysis enabled when rebased on `master`.

**Comment by MarcoFalke** [comment]
@practicalswift I think clang doesn't understand "recursive" annotations. I am happy to be proven wrong, but I think this just doesn't compile with annotations for now.

File: src/txmempool.h [hunk 4]
```diff
@@ -575,6 +575,9 @@ class CTxMemPool
     /** Translate a set of hashes into a set of pool iterators to avoid repeated lookups */
     setEntries GetIterSet(const std::set<uint256>& hashes) const EXCLUSIVE_LOCKS_REQUIRED(cs);
 
+    /** Dereference the txiter */
+    static const CTxMemPoolEntry& GetEntry(txiter it) { return *it; };
```
Comment (practicalswift) [inline src/txmempool.h]:
```
2018-09-22 21:49:48 cpplint(pr=13804): src/txmempool.h:579:  You don't need a ; after a }  [readability/braces] [4]
```

File: src/txmempool.h [hunk 5]
```diff
@@ -699,6 +713,115 @@ class CTxMemPool
...
+    public:
+        /**
+         * The lock of the underlying tx pool.
+         * The caller must acquire this lock for the whole life-time of this layer.
+         */
+        CCriticalSection& cs;
+
+        struct txiter_nested {
+            CTxMemPool::txiter _i;
+            explicit txiter_nested(CTxMemPool::txiter val) : _i(val){};
```
Comment (practicalswift) [inline src/txmempool.h]:
```
2018-09-22 21:49:48 cpplint(pr=13804): src/txmempool.h:729:  You don't need a ; after a }  [readability/braces] [4]
```

File: src/txmempool.h [hunk 6]
```diff
@@ -699,6 +713,115 @@ class CTxMemPool
...
+            txiter_nested() : _i(){};
```
Comment (practicalswift) [inline src/txmempool.h]:
```
2018-09-22 21:49:48 cpplint(pr=13804): src/txmempool.h:730:  You don't need a ; after a }  [readability/braces] [4]
```

File: src/txmempool.h [hunk 7]
```diff
@@ -699,6 +713,115 @@ class CTxMemPool
...
+        struct TxLinks {
+            CTxMemPool::Layer::setEntries parents;
+            CTxMemPool::Layer::setEntries children;
+        };
+        using txlinksMap = std::map<CTxMemPool::Layer::txiter, CTxMemPool::Layer::TxLinks, CompareIteratorByHash>;
+        /**
+         * Construct a Layer based on an existing tx pool.
+         * The caller must acquire the lock for the whole life-time of this layer.
+         */
+        Layer(const CTxMemPool& p) EXCLUSIVE_LOCKS_REQUIRED(m_tx_pool.cs)
```
Comment (practicalswift) [inline src/txmempool.h]:
```
2018-09-22 21:49:48 cpplint(pr=13804): src/txmempool.h:759:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
```

File: src/txmempool.h [hunk 8]
```diff
@@ -699,6 +713,115 @@ class CTxMemPool
...
+        void UpdateParent(const CTxMemPool::Layer::txiter& entry, const CTxMemPool::Layer::txiter& parent, bool add);
+        void UpdateChild(const CTxMemPool::Layer::txiter& entry, const CTxMemPool::Layer::txiter& child, bool add);
+        void UpdateEntryForAncestors(const CTxMemPool::Layer::txiter& it, const setEntries& setAncestors);
+        void UpdateForRemoveFromMempool(const setEntries& entriesToRemove, bool updateDescendants);
+        void UpdateChildrenForRemoval(txiter it);
+        const setEntries& GetMemPoolChildren(txiter entry) const;
+        void removeUnchecked(txiter it, MemPoolRemovalReason reason);
+
+    private:
+
```
Comment (practicalswift) [inline src/txmempool.h]:
```
2018-09-22 21:49:48 cpplint(pr=13804): src/txmempool.h:808:  Do not leave a blank line after "private:"  [whitespace/blank_line] [3]
```

File: src/txmempool.cpp [hunk 9]
```diff
@@ -389,26 +438,52 @@ void CTxMemPool::addUnchecked(const CTxMemPoolEntry &entry, setEntries &setAnces
...
+    if (minerPolicyEstimator) {
+        minerPolicyEstimator->processTransaction(entry, validFeeEstimate);
+    }
 
-    vTxHashes.emplace_back(tx.GetWitnessHash(), newit);
+    vTxHashes.emplace_back(entry.GetTx().GetWitnessHash(), newit);
     newit->vTxHashesIdx = vTxHashes.size() - 1;
 }
 
+void CTxMemPool::Layer::addUnchecked(const CTxMemPoolEntry& entry, const CTxMemPool::Layer::setEntries& setAncestors, bool /* validFeeEstimate */ ignore)
```
Comment (practicalswift) [inline src/txmempool.cpp]: Use an unnamed parameter instead of `ignore`.

File: src/txmempool.cpp [hunk 10]
```diff
@@ -389,26 +438,52 @@ void CTxMemPool::addUnchecked(const CTxMemPoolEntry &entry, setEntries &setAnces
...
+{
+    ::addUnchecked(*this, entry, setAncestors);
+}
+
+template <typename P>
+static void removeUnchecked(P& pool, typename P::txiter it, MemPoolRemovalReason reason)
```
Comment (practicalswift) [inline src/txmempool.cpp]: Make `MemPoolRemovalReason` an unnamed parameter.

File: src/txmempool.cpp [hunk 11]
```diff
@@ -804,6 +898,17 @@ CTransactionRef CTxMemPool::get(const uint256& hash) const
...
 }
 
+CTransactionRef CTxMemPool::Layer::get(const uint256& hash) const
+{
+    auto it = m_cache_added.find(hash);
+    if (it == m_cache_added.end())
+        return nullptr;
+    else
+        return it->GetSharedTx();
+    if (m_cache_removed.find(hash) != m_cache_removed.end()) return nullptr;
```
Comment (practicalswift) [inline src/txmempool.cpp]: This code will never get executed.

File: src/txmempool.cpp [hunk 12]
```diff
@@ -937,14 +1136,31 @@ int CTxMemPool::Expire(int64_t time) {
     RemoveStaged(stage, false, MemPoolRemovalReason::EXPIRY);
     return stage.size();
 }
+int CTxMemPool::Layer::Expire(int64_t time)
```
Comment (practicalswift) [inline src/txmempool.cpp]: Make `time` an unnamed parameter.

File: src/txmempool.cpp [hunk 13]
```diff
@@ -1058,6 +1316,13 @@ void CTxMemPool::TrimToSize(size_t sizelimit, std::vector<COutPoint>* pvNoSpends
     }
 }
 
+void CTxMemPool::Layer::TrimToSize(size_t sizelimit, std::vector<COutPoint>* pvNoSpendsRemaining)
```
Comment (practicalswift) [inline src/txmempool.cpp]: Make `sizelimit` an unnamed parameter.

**Pull request closed without merging**